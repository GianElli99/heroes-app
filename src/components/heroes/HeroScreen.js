import React, { useMemo } from 'react';
import { Redirect, useParams } from 'react-router-dom';
import { getHeroById } from '../../selectors/getHeroesById';

const heroesFolder = require.context('../../assets/heroes', true);

export const HeroScreen = ({ history }) => {
  const { heroId } = useParams();
  const hero = useMemo(() => getHeroById(heroId), [heroId]);

  if (!hero) {
    return <Redirect to="/" />;
  }
  const { superhero, alter_ego, publisher, first_appearance, characters } =
    hero;

  const handleReturn = () => {
    if (history.length <= 2) {
      history.push('/');
    } else {
      history.goBack();
    }
  };

  return (
    <div className="row mt-5 mx-0">
      <div className="col-4">
        <img
          src={heroesFolder(`./${heroId}.jpg`).default}
          alt="hero"
          className="img-thumbnail animate__animated animate__fadeInLeft"
        />
      </div>
      <div className="col-8 animate__animated animate__fadeIn">
        <h3>{superhero}</h3>
        <ul className="list-group list-group-flush">
          <li className="list-group-item">
            <b>Alter ego: </b>
            {alter_ego}
          </li>
          <li className="list-group-item">
            <b>Publisher: </b>
            {publisher}
          </li>
          <li className="list-group-item">
            <b>First appearance: </b>
            {first_appearance}
          </li>
        </ul>
        <h5>Characters</h5>
        <p>{characters}</p>

        <button className="btn btn-outline-info" onClick={handleReturn}>
          Return
        </button>
      </div>
    </div>
  );
};
