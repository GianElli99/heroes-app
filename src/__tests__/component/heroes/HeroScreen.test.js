import { mount, shallow } from 'enzyme';
import { MemoryRouter, Route } from 'react-router-dom';
import { HeroScreen } from '../../../components/heroes/HeroScreen';

describe('Pruebas en HeroScreen', () => {
  let historyMock = {
    push: jest.fn(),
    replace: jest.fn(),
    goBack: jest.fn(),
    length: 5,
  };

  test('should mostrar el redirect si no hay argumentos en el URL', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/hero']}>
        <HeroScreen history={historyMock} />
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();
  });

  test('debe mostrar un heroe si esta existe', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/hero/marvel-spider']}>
        <Route path="/hero/:heroId" component={HeroScreen} />
      </MemoryRouter>
    );

    expect(wrapper.find('.row').exists()).toBe(true);
  });

  test('debe regresar a la pantalla anterior con push', () => {
    historyMock.length = 1;

    const wrapper = mount(
      <MemoryRouter initialEntries={['/hero/marvel-spider']}>
        <Route
          path="/hero/:heroId"
          component={(props) => <HeroScreen history={historyMock} />}
        />
      </MemoryRouter>
    );

    wrapper.find('button').simulate('click', {});
    expect(historyMock.push).toHaveBeenCalledWith('/');
    expect(historyMock.goBack).not.toHaveBeenCalled();
  });

  test('debe regresar a la pantalla anterior', () => {
    historyMock.length = 10;

    const wrapper = mount(
      <MemoryRouter initialEntries={['/hero/marvel-spider']}>
        <Route
          path="/hero/:heroId"
          component={(props) => <HeroScreen history={historyMock} />}
        />
      </MemoryRouter>
    );

    wrapper.find('button').simulate('click', {});
    expect(historyMock.push).not.toHaveBeenCalled();
    expect(historyMock.goBack).toHaveBeenCalled();
  });

  test('debe llamar el redirect si el hero no existe', () => {
    historyMock.length = 1;

    const wrapper = mount(
      <MemoryRouter initialEntries={['/hero/marvel-spider22222222']}>
        <Route
          path="/hero/:heroId"
          component={(props) => <HeroScreen history={historyMock} />}
        />
      </MemoryRouter>
    );

    expect(wrapper.find('.row').exists()).toBe(false);
  });
});
