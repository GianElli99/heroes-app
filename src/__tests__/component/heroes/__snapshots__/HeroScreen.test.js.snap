// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en HeroScreen should mostrar el redirect si no hay argumentos en el URL 1`] = `""`;
